File: haskell/p23.hs
-- P23 (**) Extract a given number of randomly selected elements from a list.

import System.Random

rndSelect :: [x] -> Int -> [IO x]
rndSelect _ 0 = []
rndSelect l n = [randomRIO (0, length l - 1) >>= return . (l !!)] ++ (rndSelect l (n - 1))

-- Convert an IO to Int
-- (foo produces an IO Int, and bar takes an Int as its only parameter and returns a non-IO value)

-- foo >>= return . bar

-- do
--   a <- foo
--   return $ bar a

File: ben/prolog/P01-P28-lists/p23.pl
% Extract a given number of randomly selected elements from a list.

rnd_select(_, 0, []).
rnd_select(L, N0, [X|R]) :-
    N0 > 0, 
    N1 is N0 - 1,
    length(L, Len), 
    random(0, Len, I),
    nth0(I, L, X),
    rnd_select(L, N1, R).
